tiledb/sm/c_api/tiledb_experimental.h

```diff
@@ -212,6 +212,41 @@ TILEDB_EXPORT capi_return_t tiledb_array_schema_evolution_add_enumeration(
     tiledb_array_schema_evolution_t* array_schema_evolution,
     tiledb_enumeration_t* enumeration) TILEDB_NOEXCEPT;
 
+/**
+ * Extends an enumeration during array schema evolution.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_enumeration_t* original_enmr = get_existing_enumeration();
+ * const void* data = get_new_data();
+ * uint64_t data_size = get_new_data_size();
+ * tiledb_enumeration_t* new_enmr;
+ * tiledb_enumeration_extend(
+ *     ctx,
+ *     original_enmr,
+ *     data,
+ *     data_size,
+ *     nullptr,
+ *     0,
+ *     &new_enmr);
+ * tiledb_array_schema_evolution_extend_enumeration(
+ *     ctx,
+ *     array_schema_evolution,
+ *     enmr);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param array_schema_evolution The schema evolution.
+ * @param enumeration The enumeration to be extended. This should be the result
+ *        of a call to tiledb_enumeration_extend.
```

Comment:
Requiring the enumeration to come from a specific source might make this API confusing to use. How about we don't add `tiledb_enumeration_extend` and merge it into `tiledb_array_schema_evolution_extend_enumeration` like this?

```cpp
TILEDB_EXPORT capi_return_t tiledb_array_schema_evolution_extend_enumeration(
    tiledb_ctx_t* ctx,
    tiledb_array_schema_evolution_t* array_schema_evolution,
    const char* name,
    const void* data,
    uint64_t data_size,
    const void* offsets,
    uint64_t offsets_size) TILEDB_NOEXCEPT;
```

We could also rename `extend_enumeration` to `add_enumeration_members`.

Reply:
So the first thing is that its not required, its just a tip on how to easily create an extended enumeration. You could just as easily build one from scratch as long as you ensure that the data and offsets are byte-for-byte prefixes. So at the very least I'll need to rewrite that API doc.

The initial reason I went with this approach is due to lazy loading though in the end I came up against the cold reality that is our C API's exposure of the array schema evolution implementation which in the end forced me to ensure that the relevant enumerations are loaded before the evolution is applied. So technically, we could go back and add this sort of API but in hind sight there are two reasons to avoid it.

The first reason I'd prefer to avoid this API is that it'll require a lot more changes to serialization. Right now we're passing the entire enumeration across the wire so it was easy to just re-use the existing enumeration code that's already tested. Adding these API's means I'm adding a whole new message type which is kind of a bear. Its not the end of the world if we decide its worth it, but I'd rather avoid it.

The second more important reason though is the validation of the extension itself. It turns out there's a surprising number of ways that specifying those two buffers can break an extension. A lot of the funky ones are related to fixed size with cell_var_num > 1, but its enough that I think having the two steps process is worth it so that errors from extension are orthogonal from errors during evolution.

As to renaming `extend` to `add_members`, I'd prefer to stick with `extend`. The reason for that is because Enumerations are always `const` (because ArraySchema is often `const`) which means that all modifications of an Enumeration return a copy with the change rather than doing an in-place update. I know its a bit wishy-washy, but `extend` returning a copy seems more reasonable than `add_members` returning a copy instead of doing an in-place mutation.